setuptools/command/easy_install.py

```diff
@@ -1579,50 +1579,73 @@ def get_exe_prefixes(exe_filename):
-                if not path.strip() or path.strip().startswith('#'):
-                    continue
-                # skip non-existent paths, in case somebody deleted a package
-                # manually, and duplicate paths as well
-                path = self.paths[-1] = normalize_path(
-                    os.path.join(self.basedir, path)
-                )
-                if not os.path.exists(path) or path in seen:
-                    self.paths.pop()  # skip it
-                    self.dirty = True  # we cleaned up, so we're dirty now :)
-                    continue
-                seen[path] = 1
-            f.close()
+        f = open(self.filename, 'rt')
+        for line in f:
+            path = line.rstrip()
+            stripped_path = path.strip()
+            paths.append(path)  # still keep imports and empty/commented lines for formatting
+            if line.startswith(('import ', 'from ')):
+                saw_import = True
+                continue
+            if not stripped_path or stripped_path.startswith('#'):
+                continue
+            # skip non-existent paths, in case somebody deleted a package
+            # manually, and duplicate paths as well
+            normalized_path = normalize_path(os.path.join(self.basedir, path))
+            if not os.path.exists(normalized_path) or normalized_path in seen:
+                log.debug("cleaned up dirty or duplicated %r", path)
+                dirty = True
+                paths.pop()
+                continue
+            seen[normalized_path] = 1
+        f.close()
+        return paths, dirty or (paths and saw_import)
 
-        if self.paths and not saw_import:
-            self.dirty = True  # ensure anything we touch has import wrappers
-        while self.paths and not self.paths[-1].strip():
-            self.paths.pop()
+    def _load(self):
+        if os.path.isfile(self.filename):
+            paths, dirty = self._load_raw()
+            # lights cleaning:
+            while paths and not paths[-1].strip():
+                paths.pop()
+        else:
+            paths = []
+            dirty = False
+        return paths, dirty
 
     def save(self):
         """Write changed .pth file back to disk"""
+        last_paths, last_dirty = self._load()
+        for path in last_paths[:]:
+            if path not in self.paths:
+                self.paths.append(path)
+                log.info("detected new path %r", path)
+                last_dirty = True
+            else:
+                last_paths.remove(path)
+        for path in self.paths[:]:
+            if path not in last_paths \
+                    and not path.startswith(('import ', 'from ', '#')):
+                absolute_path = os.path.join(self.basedir, path)
+                if not os.path.exists(absolute_path):
+                    self.paths.remove(path)
+                    log.info("removing now non-existent path %r", path)
+                    last_dirty = True
```

Comment:
This 2 loops look tricky to me. Is there anyway we can use `set` logic (e.g. [`set.intersection`](https://docs.python.org/3/library/stdtypes.html?highlight=intersection#frozenset.intersection) and [`set.difference`](https://docs.python.org/3/library/stdtypes.html?highlight=difference#frozenset.difference)) to detect new and removed paths?

Reply:
You could use `set` to find out which ones are removed/news and then `list.remove` then from the original variable?

Reply:
looks a bit complicated compared to just looping over the refreshed view of the file and then after on the current one (minus the ones that have been already been processed). IMHO.

Reply:
I mean: I made it to be simple ;)

Reply:
the 2 loops are achieving 2 different things.